nft_ingester/src/backfiller.rs

```diff
@@ -0,0 +1,702 @@
+            )
+            .all(&self.db)
+            .await?;
+
+        Ok(rows.last().map(|row| row.seq))
+    }
+
+    async fn clear_force_chk_flag(&self, tree: &[u8]) -> Result<UpdateResult, DbErr> {
+        backfill_items::Entity::update_many()
+            .col_expr(backfill_items::Column::ForceChk, Expr::value(false))
+            .filter(backfill_items::Column::Tree.eq(tree))
+            .exec(&self.db)
+            .await
+    }
+
+    // Similar to `fetchAndPlugGaps()` in `backfiller.ts`.
+    async fn fetch_and_plug_gaps(&mut self, tree: &[u8]) -> Result<Option<i64>, IngesterError> {
+        let (opt_max_seq, gaps) = self.get_missing_data(tree).await?;
+
+        // Similar to `plugGapsBatched()` in `backfiller.ts` (although not batched).
+        for gap in gaps.iter() {
+            // Similar to `plugGaps()` in `backfiller.ts`.
+            let _result = self.plug_gap(&gap, tree).await?;
+        }
+
+        Ok(opt_max_seq)
+    }
+
+    // Similar to `getMissingData()` in `db.ts`.
+    async fn get_missing_data(&self, tree: &[u8]) -> Result<(Option<i64>, Vec<GapInfo>), DbErr> {
+        // Get the maximum sequence number that has been backfilled, and use
+        // that for the starting sequence number for backfilling.
+        let mut query = backfill_items::Entity::find()
+            .select_only()
+            .column(backfill_items::Column::Seq)
+            .filter(backfill_items::Column::Backfilled.eq(true))
+            .filter(backfill_items::Column::Tree.eq(tree))
+            .group_by(backfill_items::Column::Tree)
+            .build(DbBackend::Postgres);
+
+        query.sql = query.sql.replace(
+            "SELECT \"backfill_items\".\"seq\"",
+            "SELECT MAX (\"backfill_items\".\"seq\")",
+        );
+
+        // Debug.
+        //println!("{}", query.to_string());
+
+        let start_seq_vec = MaxSeqItem::find_by_statement(query).all(&self.db).await?;
+        let start_seq = if start_seq_vec.len() > 0 {
+            start_seq_vec[0].max
+        } else {
+            0
+        };
+
+        // Debug.
+        //println!("MAX: {}", start_seq);
+
+        // Get all rows for the tree that have not yet been backfilled.
+        let mut query = backfill_items::Entity::find()
+            .select_only()
+            .column(backfill_items::Column::Seq)
+            .column(backfill_items::Column::Slot)
+            .filter(backfill_items::Column::Seq.gte(start_seq))
+            .filter(backfill_items::Column::Tree.eq(tree))
+            .order_by_asc(backfill_items::Column::Seq)
+            .build(DbBackend::Postgres);
+
+        query.sql = query.sql.replace("SELECT", "SELECT DISTINCT");
+        let rows = SimpleBackfillItem::find_by_statement(query)
+            .all(&self.db)
+            .await?;
+        let mut gaps = vec![];
+
+        // Look at each pair of subsequent rows, looking for a gap in sequence number.
+        for (prev, curr) in rows.iter().zip(rows.iter().skip(1)) {
```

Comment:
> At a minimum this is O(n) time complexity

Yes n here is the number of items in the `backfill_items` that had multiple rows for a given tree.  Any time a Bubblegum instruction is detected it will add one row to `backfill_items` and then the backfiller should check, so if it is super caught up then N would be 2.  I agree in general N is number of non-deleted rows that haven't been run yet, backfiller could get behind and N goes up if contract is slammed.

> Does the zip add alot of overhead

I don't think so, Rust has goals for these to be _zero-cost abstraction_, but below is a small test and results to see for ourselves:
```
use rand::Rng;

fn main() {
    let mut rng = rand::thread_rng();
    let size: u64 = rng.gen_range(1..10_000_000);
    println!("Initial of test: {size}");

    println!("Gen test vals");
    let mut list: Vec<u64> = (0..size).collect();

    let num_to_remove = rng.gen_range(0..std::cmp::min(size, 100_000));
    for _ in 0..num_to_remove {
        list.swap_remove(rng.gen_range(0..list.len()));
    }
    list.sort();
    println!("Test vals len after random remove {}", list.len());

    let mut out_vec_1 = vec![];
    let mut out_vec_2 = vec![];

    let started = std::time::Instant::now();

    for i in 0..(list.len() - 1) {
        if list[i + 1] - list[i] > 1 {
            out_vec_1.push((list[i].clone(), list[i + 1].clone()));
        }
    }
    println!("Indexed version: {:?}", started.elapsed());

    let started = std::time::Instant::now();
    for (v1, v2) in list.iter().zip(list.iter().skip(1)) {
        if v2 - v1 > 1 {
            out_vec_2.push((v1.clone(), v2.clone()));
        }
    }
    println!("Zip version: {:?}", started.elapsed());

    println!("Final check");
    let count = out_vec_1
        .iter()
        .zip(&out_vec_2)
        .filter(|(pair_1, pair_2)| *pair_1 == *pair_2)
        .count();

    println!(
        "Vec compare count {} {} {}",
        out_vec_1.len(),
        out_vec_2.len(),
        count
    );
}
```
### Results:
```
danenbm@Michaels-MacBook-Pro zip_testing % cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
     Running `target/debug/zip_testing`
Initial of test: 891122
Gen test vals
Test vals len after random remove 889309
Indexed version: 37.94075ms
Zip version: 31.733ms
Final check
Vec compare count 1806 1806 1806

danenbm@Michaels-MacBook-Pro zip_testing % cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
     Running `target/debug/zip_testing`
Initial of test: 281351
Gen test vals
Test vals len after random remove 202581
Indexed version: 11.925125ms
Zip version: 8.640583ms
Final check
Vec compare count 56788 56788 56788

danenbm@Michaels-MacBook-Pro zip_testing % cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
     Running `target/debug/zip_testing`
Initial of test: 6947780
Gen test vals
Test vals len after random remove 6877138
Indexed version: 296.892166ms
Zip version: 245.677291ms
Final check
Vec compare count 69965 69965 69965
```